src/file-events/cpp/win_fsnotifier.cpp

```diff
@@ -8,14 +10,25 @@ using namespace std;
 // WatchPoint
 //
 
-WatchPoint::WatchPoint(Server* server, const u16string& path, HANDLE directoryHandle) {
+static void CALLBACK startWatchCallback(_In_ ULONG_PTR arg) {
+    WatchPoint* watchPoint = (WatchPoint*) arg;
+    watchPoint->listen();
+}
+
+WatchPoint::WatchPoint(Server* server, const u16string& path, HANDLE directoryHandle, HANDLE serverThreadHandle) {
     this->server = server;
     this->path = path;
     this->buffer = (FILE_NOTIFY_INFORMATION*) malloc(EVENT_BUFFER_SIZE);
     ZeroMemory(&this->overlapped, sizeof(OVERLAPPED));
     this->overlapped.hEvent = this;
     this->directoryHandle = directoryHandle;
     this->status = WATCH_UNINITIALIZED;
+    unique_lock<mutex> lock(listenerMutex);
+    QueueUserAPC(startWatchCallback, serverThreadHandle, (ULONG_PTR) this);
+    listenerStarted.wait(lock);
+    if (status != WATCH_LISTENING) {
+        throw new FileWatcherException("Couldn't start listening");
```

Comment:
Can we add the path we try to watch to this exception?

Reply:
No. Dynamic strings are not a thing in C++ exceptions unfortunately. :/

Reply:
:(